controllers/state_of_the_world.go

```diff
@@ -116,33 +194,92 @@ func NewPolicyMachineryController(manager ctrlruntime.Manager, client *dynamic.D
 	if err != nil || !ok {
 		logger.Info("cert manager is not installed, skipping related watches and reconcilers", "err", err)
 	} else {
-		controllerOpts = append(controllerOpts,
-			controller.WithRunnable("certificate watcher", controller.Watch(&certmanagerv1.Certificate{}, CertManagerCertificatesResource, metav1.NamespaceAll)),
-			controller.WithRunnable("issuers watcher", controller.Watch(&certmanagerv1.Issuer{}, CertManagerIssuersResource, metav1.NamespaceAll)),
-			controller.WithRunnable("clusterissuers watcher", controller.Watch(&certmanagerv1.Certificate{}, CertMangerClusterIssuersResource, metav1.NamespaceAll)),
-			controller.WithObjectKinds(
-				CertManagerCertificateKind,
-				CertManagerIssuerKind,
-				CertManagerClusterIssuerKind,
-			),
-			// TODO: add object links
-		)
-		// TODO: add tls policy specific tasks to workflow
+		controllerOpts = append(controllerOpts, certManagerControllerOpts()...)
 	}
 
+	controllerOpts = append(controllerOpts, controller.WithReconcile(buildReconciler(client, manager.GetRESTMapper())))
+
 	return controller.NewController(controllerOpts...)
 }
 
-func buildReconciler(client *dynamic.DynamicClient) controller.ReconcileFunc {
+func buildReconciler(client *dynamic.DynamicClient, restMapper meta.RESTMapper) controller.ReconcileFunc {
 	reconciler := &controller.Workflow{
-		Precondition: NewEventLogger().Log,
+		Precondition: (&controller.Workflow{
+			Precondition: NewEventLogger().Log,
+			Tasks: []controller.ReconcileFunc{
+				NewTopologyFileReconciler(client, operatorNamespace).Reconcile,
+				NewIsCertManagerInstalledTask(restMapper).Reconcile,
+			},
+		}).Run,
 		Tasks: []controller.ReconcileFunc{
-			NewTopologyFileReconciler(client, operatorNamespace).Reconcile,
+			NewTLSPolicyWorkflow(client).Run,
 		},
 	}
 	return reconciler.Run
 }
 
+func certManagerControllerOpts() []controller.ControllerOption {
+	isCertificateOwnedByTLSPolicy := func(c *certmanagerv1.Certificate) bool {
+		return isObjectOwnedByKind(c, kuadrantv1alpha1.TLSPolicyKind.Group, kuadrantv1alpha1.TLSPolicyKind.Kind)
+	}
+
+	return []controller.ControllerOption{
+		controller.WithRunnable("certificate watcher", controller.Watch(
+			&certmanagerv1.Certificate{},
+			CertManagerCertificatesResource,
+			metav1.NamespaceAll,
+			controller.WithPredicates(ctrlruntimepredicate.TypedFuncs[*certmanagerv1.Certificate]{
+				CreateFunc: func(e event.TypedCreateEvent[*certmanagerv1.Certificate]) bool {
+					return isCertificateOwnedByTLSPolicy(e.Object)
+				},
+				UpdateFunc: func(e event.TypedUpdateEvent[*certmanagerv1.Certificate]) bool {
+					return isCertificateOwnedByTLSPolicy(e.ObjectNew)
+				},
+				DeleteFunc: func(e event.TypedDeleteEvent[*certmanagerv1.Certificate]) bool {
+					return isCertificateOwnedByTLSPolicy(e.Object)
+				},
+				GenericFunc: func(e event.TypedGenericEvent[*certmanagerv1.Certificate]) bool {
+					return isCertificateOwnedByTLSPolicy(e.Object)
+				},
+			})),
```

Comment:
Interesting approach. I wonder how it performs compared to labels.

Reply:
Yeah, I went with ownerReference instead since we already set the ownerRef to point to `TLSPolicy` since we want the associated `Certificates` deleted when the `TLSPolicy` is deleted.

Also we don't currently set a common label across all created `Certificate`'s
https://github.com/Kuadrant/kuadrant-operator/blob/4793dc4d2ad522a96f696c68f448e8b4518a6147/controllers/tlspolicy_certmanager_certificates.go#L161-L184